phac-nml/irida-next
# Add graphql support for samples via sample_type and project_samples_resolver

## What does this PR do and why?
Adds graphql support for samples.
Allows samples to be fetched via nodes (gid)
Allows samples to be fetched through projects

Added SampleType to graphql schema for Sample objects
SamplesResolver handles authorization scope for samples
ProjectSamplesResolver handles scoping samples to project

## Screenshots or screen recordings
This PR is for graphql, no UI changes

## How to set up and validate locally
1. Setup, do the usual
```
//pull branch
bundle
rake db:drop db:create db:migrate db:seed
//run 
```
2. Verify queries are working as intended

sign in with admin@email.com to authorize
go to http://localhost:3000/graphiql

Create the following queries to test
```
query viewOneProject {
  node(id: "gid://irida/Project/2") {
    ... on Project {
      fullName
      fullPath
      id
      name
      path
      samples {
        nodes {
          id
          name
          projectId
        }
      }
    }
  }
}

query getSamples {
  samples {
    nodes {
      id
    }
  }
}

query viewOneSample {
  node(id: "gid://irida/Sample/11") {
    ... on Sample {
      name
      description
      id
      projectId
    }
  }
}
```
2.1. `viewOneProject` will query project 2 and list the samples as nodes
```
{
  "data": {
    "node": {
      "fullName": "Bacillus / Bacillus anthracis / Outbreak 2022",
      "fullPath": "bacillus/bacillus-anthracis/outbreak-2022",
      "id": "gid://irida/Project/2",
      "name": "Outbreak 2022",
      "path": "outbreak-2022",
      "samples": {
        "nodes": [
          {
            "id": "gid://irida/Sample/11",
            "name": "Bacillus anthracis/Outbreak 2022 Sample 1",
            "projectId": "2"
          },
          {
            "id": "gid://irida/Sample/12",
            "name": "Bacillus anthracis/Outbreak 2022 Sample 2",
            "projectId": "2"
          },
     // more samples here
          {
            "id": "gid://irida/Sample/20",
            "name": "Bacillus anthracis/Outbreak 2022 Sample 10",
            "projectId": "2"
          }
        ]
      }
    }
  }
}
```
2.2. `getSamples` will get every sample that the user has access to.
```
{
  "data": {
    "samples": {
      "nodes": [
        {
          "id": "gid://irida/Sample/1"
        },
        {
          "id": "gid://irida/Sample/2"
        },
   // more samples here
        {
          "id": "gid://irida/Sample/340"
        }
      ]
    }
  }
}
```
2.2.1. To test this further, you can 
a. create a new user,
b. create a project,
c. run the query and see that no samples are returned
```
{
  "data": {
    "samples": {
      "nodes": []
    }
  }
}
```
d. add some samples to the project
e. run the query and see that only these newly created samples are shown to user
```
{
  "data": {
    "samples": {
      "nodes": [
        {
          "id": "gid://irida/Sample/1421"
        },
        {
          "id": "gid://irida/Sample/1422"
        }
      ]
    }
  }
}
```

2.3. `viewOneSample` fetching a single sample based on the `gid`
//while signed in to admin@email.com
```
{
  "data": {
    "node": {
      "name": "Bacillus anthracis/Outbreak 2022 Sample 1",
      "description": "This is a description for sample Bacillus anthracis/Outbreak 2022 Sample 1.",
      "id": "gid://irida/Sample/11",
      "projectId": "2"
    }
  }
}
```
2.3.1. to test this further, you can
a. sign in with a new user
b. run the query and see that the user is not authorized
```
{
  "data": {
    "node": null
  },
  "errors": [
    {
      "message": "An object of type Sample was hidden due to permissions",
      "locations": [
        {
          "line": 38,
          "column": 3
        }
      ],
      "path": [
        "node"
      ]
    }
  ]
}
```

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
